### PR TITLE
libtiff_ctype: On Windows, use TIFFOpenW() when open() gets a str

### PR DESCRIPTION
**Pull request by pieleric**
PR #160 improved the handling of non-ascii character in filenames on Windows. However, it doesn't work in every case. On my Windows version, the filesystem encoding is reported as cp1252 (ie, latin), and still, I can have files with Chinese (or non latin) characters! When opening such file, it still fails because the path cannot be encoded to "cp1252".

So, I'm suggesting to change the logic, and just use the dedicated TIFFOpenW() function which takes as argument a unicode string... and seems to magically work in every case.

Please note that this current PR now changes the code to separate the Python 2 and Python 3 support (because "str" means different things on Python 2 & 3). However, I'm not sure whether it's intentional to still support Python 2. So if you prefer, I could just drop the Python 2 part.

### CODE REVIEW

**Comment by djhoese** [comment]
I am just a small contributor/maintainer to this project, but my vote is to drop Python 2 (actually drop Python <=3.9). The final decision is @pearu's though.

**Comment by pearu** [comment]
According to setup.py, pylibtiff already requires Python >= 3.8.

**Comment by djhoese** [comment]
Ah I must have done that in a recent update. Thanks. @pieleric I guess that means any Python 2 support you see can be removed.

**Comment by pieleric** [comment]
I've just updated the PR by completely dropping the Python 2 path. It nicely simplifies the code :smile:

File: libtiff/libtiff_ctypes.py
Comment (djhoese) [inline libtiff/libtiff_ctypes.py:545]: I know this was `Exception` before, but any idea if we can make this more specific?

File: libtiff/libtiff_ctypes.py
Comment (djhoese) [inline libtiff/libtiff_ctypes.py:548]: Personal preference:

```suggestion
                print(f"Warning: filename argument is of wrong type or encoding: {ex}")
```

Comment (djhoese) [reply libtiff/libtiff_ctypes.py:548]: Or... @pearu thoughts on `warnings.warn`?

Comment (pearu) [reply libtiff/libtiff_ctypes.py:548]: Yes, using `warnings.warn` is preferable.

File: libtiff/libtiff_ctypes.py
Comment (pearu) [inline libtiff/libtiff_ctypes.py:530]: Can we stick to `str` only? That is, in `open`, the filename needs to be encoded to `bytes` but is there are use case where callers of `open` would prefer giving the `filename` as `bytes` object?

File: libtiff/libtiff_ctypes.py
Comment (pieleric) [reply libtiff/libtiff_ctypes.py:548]: I'll happily execute... Honestly, I'd love removing all the `print()`. They should either to logging or to warnings.warn (or even to an Exception).
Our software is a GUI, and when using this library, all the print's are essentially sent to nowhere. Using logging is much easier to redirect them to a log file.

File: libtiff/libtiff_ctypes.py
Comment (pieleric) [reply libtiff/libtiff_ctypes.py:530]: In our case, we always use `str` to encode the path, as a convention in our application. However, so far in Python the convention is to accept both str and bytes for paths (like the argument to `open()`). It's called a "path-like object": https://docs.python.org/3/glossary.html#term-path-like-object
So I'd suggest to follow the same convention, and give freedom to the user.

File: libtiff/libtiff_ctypes.py
Comment (pieleric) [reply libtiff/libtiff_ctypes.py:545]: Indeed, I just used the original code. I'm pretty sure it should be safe to just catch `UnicodeError`. At least that's what the warning afterwards assumes is the issue.

File: libtiff/libtiff_ctypes.py
Comment (pieleric) [reply libtiff/libtiff_ctypes.py]: I've now changed the code to call `os.fspath()`, so that it should work with any path-like object (eg, also the PathLib).

File: libtiff/libtiff_ctypes.py
Comment (pieleric) [reply libtiff/libtiff_ctypes.py]: Sorry, that was carried over from the old code, which also supported Python 2 (where it needs "str"). I've updated that comment.

File: libtiff/libtiff_ctypes.py
Comment (pearu) [inline libtiff/libtiff_ctypes.py:540]: Does it make sense to always use TIFFOpenW on Windows? E.g. via
```suggestion
        if hasattr(libtiff, "TIFFOpenW"):
            if isinstance(filename, bytes):
                filename = filename.decode()
```

Comment (pieleric) [reply libtiff/libtiff_ctypes.py:540]: We'd need to know from which encoding to decode from... and be (almost) back to doing what is done in the case of `TIFFOpen()`. I think it's fair to say if you pass bytes, we call the function that uses bytes, and if you pass unicode, we call the function that uses unicode.

Comment (pearu) [reply libtiff/libtiff_ctypes.py:540]: SGTM, thanks!

**Review by pearu — APPROVED** [review]
LGTM! Thanks, @pieleric!

**Review by djhoese — APPROVED** [review]
Thanks!

**Pull request merged**
